x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/pages/rule_creation/index.tsx

```diff
@@ -328,104 +335,177 @@ const CreateRulePageComponent: React.FC = () => {
-        const startMlJobsIfNeeded = async () => {
-          if (!isMlRule(ruleType) || !enabled) {
-            return;
-          }
-          await startMlJobs(localDefineStepData.machineLearningJobId);
-        };
-        const [, createdRule] = await Promise.all([
-          startMlJobsIfNeeded(),
-          createRule(
-            formatRule<RuleCreateProps>(
-              localDefineStepData,
-              localAboutStepData,
-              localScheduleStepData,
-              {
-                ...localActionsStepData,
-                enabled,
-              },
-              triggersActionsUi.actionTypeRegistry
-            )
-          ),
-        ]);
-
-        addSuccess(i18n.SUCCESSFULLY_CREATED_RULES(createdRule.name));
-
-        navigateToApp(APP_UI_ID, {
-          deepLinkId: SecurityPageName.rules,
-          path: getRuleDetailsUrl(createdRule.id),
-        });
-      }
+  const createRuleFromFormData = useCallback(
+    async (enabled: boolean) => {
+      const localDefineStepData: DefineStepRule = defineFieldsTransform({
+        ...defineStepForm.getFormData(),
+        eqlOptions: eqlOptionsSelected,
+      });
+      const localAboutStepData = aboutStepForm.getFormData();
+      const localScheduleStepData = scheduleStepForm.getFormData();
+      const localActionsStepData = actionsStepForm.getFormData();
+      const startMlJobsIfNeeded = async () => {
+        if (!isMlRule(ruleType) || !enabled) {
+          return;
+        }
+        await startMlJobs(localDefineStepData.machineLearningJobId);
+      };
+      const [, createdRule] = await Promise.all([
+        startMlJobsIfNeeded(),
+        createRule(
+          formatRule<RuleCreateProps>(
+            localDefineStepData,
+            localAboutStepData,
+            localScheduleStepData,
+            {
+              ...localActionsStepData,
+              enabled,
+            },
+            triggersActionsUi.actionTypeRegistry
+          )
+        ),
+      ]);
+
+      addSuccess(i18n.SUCCESSFULLY_CREATED_RULES(createdRule.name));
+
+      navigateToApp(APP_UI_ID, {
+        deepLinkId: SecurityPageName.rules,
+        path: getRuleDetailsUrl(createdRule.id),
+      });
     },
     [
-      validateStep,
-      defineStepForm,
-      eqlOptionsSelected,
       aboutStepForm,
-      scheduleStepForm,
       actionsStepForm,
-      createRule,
       addSuccess,
+      createRule,
+      defineFieldsTransform,
+      defineStepForm,
+      eqlOptionsSelected,
       navigateToApp,
       ruleType,
+      scheduleStepForm,
       startMlJobs,
-      defineFieldsTransform,
       triggersActionsUi.actionTypeRegistry,
     ]
   );
 
+  const submitRule = useCallback(
+    async (enabled: boolean) => {
+      const { valid, blockingErrors, nonBlockingErrors } = await validateEachStep();
+      if (valid) {
+        // There are no validation errors, thus proceed to rule creation
+        createRuleFromFormData(enabled);
```

Comment:
this function is asynchronous. should we wait until it resolves?

Reply:
done